packages/Meshfree/src/DTK_SplineProlongationOperator.hpp

```diff
@@ -0,0 +1,168 @@
+  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+//---------------------------------------------------------------------------//
+/*!
+ * \file   DTK_SplineProlongationOperator.hpp
+ * \author Stuart R. Slattery
+ * \brief  Spline transformation operator.
+ */
+//---------------------------------------------------------------------------//
+
+#ifndef DTK_SPLINEPROLONGATIONOPERATOR_HPP
+#define DTK_SPLINEPROLONGATIONOPERATOR_HPP
+
+#include <DTK_Types.h>
+
+#include <Teuchos_Comm.hpp>
+#include <Teuchos_RCP.hpp>
+
+#include <Tpetra_Map.hpp>
+#include <Tpetra_MultiVector.hpp>
+#include <Tpetra_Operator.hpp>
+
+namespace DataTransferKit
+{
+//---------------------------------------------------------------------------//
+/*!
+ * \class SplineProlongationOperator
+ * \brief Prolongation operator for projecting a vector into the extended
+ * spline space.
+ */
+//---------------------------------------------------------------------------//
+template <typename Scalar, typename LocalOrdinal, typename GlobalOrdinal,
+          typename Node>
+class SplineProlongationOperator
+    : public Tpetra::Operator<Scalar, LocalOrdinal, GlobalOrdinal, Node>
+{
+    using Map = Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>;
+    using MultiVector =
+        Tpetra::MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node>;
+
+  public:
+    // Constructor.
+    SplineProlongationOperator( const int num_polynomial_dofs,
+                                const Teuchos::RCP<const Map> &domain_map )
+        : _domain_map( domain_map )
+    {
+        // Create a range map.
+        Teuchos::ArrayView<const GlobalOrdinal> domain_elements =
+            _domain_map->getNodeElementList();
+        _lda = domain_elements.size();
+
+        const auto old_size = domain_elements.size();
+        Teuchos::Array<GlobalOrdinal> global_ids( old_size +
+                                                  num_polynomial_dofs );
+        if ( _domain_map->getComm()->getRank() == 0 )
+        {
+            GlobalOrdinal max_id = _domain_map->getMaxAllGlobalIndex() + 1;
+
+            global_ids( 0, old_size ).assign( domain_elements );
+            for ( int i = 0; i < num_polynomial_dofs; ++i )
+            {
+                global_ids[old_size + i] = max_id + i;
```

Comment:
Why do we do this?

Reply:
We tuck in DOFs corresponding to polynomial at the end. We want to make sure they don't overlap with the dofs of the map we constructed for sources, thus we take the max dofs of that map, and add to it.